R/loo_difference_plot.R

```diff
@@ -0,0 +1,151 @@
+#' `y` rather than `y` itself. Useful when `y` has a very irregular 
+#' distribution.
+#' @param sortByGroup Sort observations by `group`, then plot against an 
+#' arbitrary index. Plotting by index can be useful when categories have 
+#' very different sample sizes. 
+#' 
+#' 
+#' @template return-ggplot
+#' 
+#' @template reference-vis-paper
+#' 
+#' @examples 
+#' 
+#' library(loo)
+#' 
+#' cbPalette <- c("#636363", "#E69F00", "#56B4E9", "#009E73", 
+#'                "#F0E442", "#0072B2","#CC79A7")
+#' 
+#' # Plot using groups from WHO
+#' 
+#' plot_loo_dif(factor(GM@data$super_region_name), loo3, loo2, 
+#'              group = GM@data$super_region_name, alpha = .5, 
+#'              jitter = c(.45, .2)
+#'              ) + 
+#'              xlab("Region") + scale_colour_manual(values=cbPalette)
+#' 
+#' # Plot using groups identified with clustering
+#' 
+#' plot_loo_dif(factor(GM@data$cluster_region), loo3, loo2, 
+#'              group = GM@data$super_region_name, alpha = .5, 
+#'              jitter = c(.45, .2)
+#'              ) + 
+#'              xlab("Cluster Group") + scale_colour_manual(values=cbPalette)
+#'              
+#' # Plot using an index variable to reduce crowding
+#' 
+#' plot_loo_dif(1:2980, loo3, loo2, group = GM@data$super_region_name, 
+#'              alpha = .5, sortByGroup = TRUE, 
+#'              ) + 
+#'              xlab("Index") + scale_colour_manual(values=cbPalette)
+#' 
+#' 
+plot_loo_variation <- 
+  function(y,
+           psis_object_1,
+           psis_object_2,
+           ...,
+           group = FALSE,
+           outlier_thresh = FALSE,
+           size = 1,
+           alpha = 1,
+           jitter = 0,
+           quantiles = FALSE,
+           sortByGroup = FALSE
+  ){
+    
+    # Adding a 0 at the end lets users provide a single number as input.
+    # In this case, only horizontal jitter is applied.
+    jitter <- c(jitter, 0) 
+    
+    elpdDif <- psis_object_1$pointwise[, "elpd_loo"] - 
+               psis_object_2$pointwise[, "elpd_loo"]
+    
+
+    if (quantiles){
+      # If quantiles is set to true, replace all y values with their quantile
+      y <- ecdf(y)(y)
+    }
+
+    
+    if (sortByGroup){
+      if (identical(group, FALSE) || !identical(y, 1:length(y))){
+        stop("ERROR: sortByGroup should only be used for grouping categorical 
+             variables, then plotting them with an arbitrary index. You can
+             create such an index using `1:length(data)`.
+             ")
+      }
+      
+      values <- group_by(tibble(group, elpdDif), factor(group)) %>%
+                    arrange(.by_group = TRUE)
+      
+      elpdDif <- pull(values, elpdDif)
+      group <- pull(values, group)
+      
+    }
+    
+    
+    plot <- ggplot(mapping=aes(y, elpdDif)) +
+            geom_hline(yintercept=0) + 
+            xlab(ifelse(sortByGroup, "y", "Index")) +
+            ylab(expression(ELPD[i][1] - ELPD[i][2])) + 
+            labs(color = "Groups")
```

Comment:
In order to call functions from ggplot (or any other package) inside of the loo package source code we need to prefix them with `ggplot2::`, e.g. `ggplot2::geom_hline()`.

Reply:
For example, here are some calls to the parallel package from a different file in the loo package: 

https://github.com/stan-dev/loo/blob/01178587e101c4e06e5ab078acdf69ab272a75f2/R/loo_moment_matching.R#L130-L133

We need to do the same thing with ggplot2.